### DISCUSSION TITLE
Can you set required fields in a request body of fields in a "oneOf" or "anyOf"

### DISCUSSION THREAD

**Discussion by JordanGardiner-enable** [OAI/OpenAPI-Specification#3113] [Q&A]
Hi,

I have been trying to work out how to set fields as required for a specific method requestBody when I am using a "oneOf" in a schema object.

I know how to set required fields of a referenced object, but when you try set the required properties of oneOf objects the same way, it does not work.

I also know that I can set the required fields in the schema object itself, however I do not want to do this as the required fields differ on each method (i.e all fields required for a post, but only some required for a patch).

So essentially, how do I set the required fields in a request body of a schema used in oneOf?

I have created an example below, in it, I want to:

-  Set the fields Bird - wingSpan and beakLength as required in the Post method 
- Set the fields Cat - tailLength as required in the Post method
- Not have any required fields for Bird and Cat in the Patch method.

```
openapi: 3.0.3
servers:
  - description: SwaggerHub API Auto Mocking
    url: https://virtserver.swaggerhub.com/Enable-Networks/TroubleTicket/1.0.0

info:
  title: Pet API
  version: 1.0.0

paths:
  /pet:
    patch:
      summary: Update a pet
      requestBody:
        required: true
        content:
          application/json:
            schema:
              allOf:
                - $ref: "#/components/schemas/Pet"
                - type: object
                  required:
                    - age
      responses:
        "204":
          description: The request was a success and the pet was successfully created.

    post:
      summary: Create a pet
      requestBody:
        required: true
        content:
          application/json:
            schema:
              allOf:
                - $ref: "#/components/schemas/Pet"
                - type: object
                  required:
                    - name
                    - age
                    - species
                  properties:
                    species:
                      type: object
                      required:
                        - ???
      responses:
        "204":
          description: The request was a success and the pet was successfully created.

components:
  schemas:
    Cat:
      type: object
      properties:
        tailLength: 
          type: integer
        whiskerLength:
          type: integer
    Bird:
      type: object 
      properties:
        wingSpan:
          type: integer
        beakLength:
          type: integer
        highestAltitude:
          type: integer

    Pet:
      type: object
      properties:
        name:
          type: string
        age:
          type: string
        species:
          oneOf:
            - $ref: '#/components/schemas/Cat'
            - $ref: '#/components/schemas/Bird'
```

Any help on this would be appreciated!

**Comment by jeremyfiel** [comment]
you might try https://join.slack.com/t/json-schema/shared_invite/zt-1rptj13ch-ZygpGVq7jUxQYpct0vwsGQ

ultimately, this is a JSON Schema behavior.

**Comment by handrews** [comment]
Closing as this is a JSON Schema question, but the short answer is that you need to put the `required` keywords under the `oneOf` branches and alongside the `$ref`s there (depending on your version of JSON Schema / OAS, that might require pushing the `$ref` into a nested `allOf`).
